scripts/cli/src/util/init.ts

```diff
@@ -0,0 +1,274 @@
+	// Should use a mutex here...
+	while (!done) {
+		await new Promise((resolve) => setTimeout(resolve, 1000));
+	}
+}
+
+// Fetches DID that belongs to the Account Key
+export async function keyToIdentityIds(
+	api: ApiPromise,
+	accountKey: AccountId | KeyringPair["publicKey"]
+): Promise<IdentityId> {
+	let account_did = await api.query.identity.keyToIdentityIds(accountKey);
+	return account_did.toHuman() as IdentityId;
+}
+
+// Returns the asset did
+export function tickerToDid(ticker: Ticker) {
+	let tickerString = String.fromCharCode.apply(ticker);
+	let tickerUintArray = Uint8Array.from(tickerString, (x) => x.charCodeAt(0));
+	return blake2AsHex(u8aConcat(stringToU8a("SECURITY_TOKEN:"), u8aFixLength(tickerUintArray, 96, true)));
+}
+
+export async function generateStashKeys(api: ApiPromise, accounts: KeyringPair[]) {
+	let keys = [];
+	await cryptoWaitReady();
+	for (let i = 0; i < accounts.length; i++) {
+		keys.push(
+			new Keyring({ type: "sr25519" }).addFromUri(`//${accounts[i]}//stash`, {
+				name: `${accounts[i] + "_stash"}`,
+			})
+		);
+		let accountRawNonce = (await api.query.system.account(keys[i].address)).nonce;
+		let account_nonce = new BN(accountRawNonce.toString());
+		nonces.set(keys[i].address, account_nonce);
+	}
+	return keys;
+}
+
+export function sendTransaction(
+	transaction: SubmittableExtrinsic<"promise">,
+	signer: KeyringPair,
+	nonceObj: NonceObject
+) {
+	return new Promise((resolve, reject) => {
+		const gettingUnsub = transaction.signAndSend(signer, nonceObj, (receipt) => {
+			const { status } = receipt;
+
+			if (receipt.isCompleted) {
+				/*
+				 * isCompleted === isFinalized || isError, which means
+				 * no further updates, so we unsubscribe
+				 */
+				gettingUnsub.then((unsub) => {
+					unsub();
+				});
+
+				if (receipt.isInBlock) {
+					// tx included in a block and finalized
+					const failed = receipt.findRecord("system", "ExtrinsicFailed");
+
+					if (failed) {
+						// get revert message from event
+						let message = "";
+						const dispatchError = (failed.event.data[0] as unknown) as DispatchError;
+
+						if (dispatchError.isModule) {
+							// known error
+							const mod = dispatchError.asModule;
+							const { section, name, documentation } = mod.registry.findMetaError(
+								new Uint8Array([mod.index.toNumber(), mod.error.toNumber()])
+							);
+
+							message = `${section}.${name}: ${documentation.join(" ")}`;
+						} else if (dispatchError.isBadOrigin) {
+							message = "Bad origin";
+						} else if (dispatchError.isCannotLookup) {
+							message = "Could not lookup information required to validate the transaction";
+						} else {
+							message = "Unknown error";
+						}
+
+						reject(new Error(message));
+					} else {
+						resolve(receipt);
+					}
+				} else if (receipt.isError) {
+					reject(new Error("Transaction Aborted"));
+				}
+			}
+		});
+	});
+}
+
+export async function handle <T>(promise: Promise<T>): Promise<T[] | [any, T]> {
```

Comment:
I don't see any needs to this handle function. It will prevent you from getting correct types when fetching data from the ApiPromise once you generate the correct types. Probably you could fetch directly from the API on every method.

Reply:
Sounds interesting can you provide an example of this, or share a link ? As I would like to be able to handle errors as efficiently as possible, thanks in advance.

Reply:
I mean you could call directly to 
```api.query.identity.authorizations.entries({
	Account: secondaryKeys[i].publicKey,
})``` in the method and get this value typed instead of loose its type by calling `handle`. 
In the typegen link I shared you can see some examples of how these queries are correctly typed: https://polkadot.js.org/docs/api/examples/promise/typegen

Reply:
I guess the part that I don't quite understand is say like in the example they provide in the like where they have 

```
// get a query
  const recordOpt = await api.query.voting.voteRecords(123);
```
in the case of a reject from the promise would it be uncaught ?

Reply:
I was looking at https://dev.to/sobiodarlington/better-error-handling-with-async-await-2e5m in terms of the error handling part where I tried to make sure no promise was uncaught. I would make the changes you suggested to handle types better.